libass/ass_render.c

```diff
@@ -2203,244 +2134,192 @@ static void render_and_combine_glyphs(ASS_Renderer *render_priv,
-                            info->bm->stride,
-                            info->bm->h,
-                            info->bm->w
-                        );
+                }
+            }
+            if (!bord && info->n_bm_o == 1) {
+                for (int j = 0; j < info->bitmap_count; ++j) {
+                    if (!info->bitmaps[j].image->bm_o)
+                        continue;
+                    info->bm_o = copy_bitmap(info->bitmaps[j].image->bm_o);
+                    if (info->bm_o) {
+                        info->bm_o->left += info->bitmaps[j].x;
+                        info->bm_o->top  += info->bitmaps[j].y;
                     }
-                    if(info->bm_o && info->bm_o->w && info->bm_o->h){
-                        offset_x = (info->pos.x >> 6) - current_info->pos.x + info->bm_o->left;
-                        offset_y = (info->pos.y >> 6) - current_info->pos.y + info->bm_o->top;
-                        render_priv->add_bitmaps_func(
-                            &current_info->bm_o->buffer[offset_y * current_info->bm_o->stride + offset_x],
-                            current_info->bm_o->stride,
-                            info->bm_o->buffer,
-                            info->bm_o->stride,
-                            info->bm_o->h,
-                            info->bm_o->w
-                        );
+                    break;
+                }
+            } else if (info->n_bm_o) {
+                info->bm_o = alloc_bitmap(info->rect_o.x_max - info->rect_o.x_min + 2 * bord,
+                                          info->rect_o.y_max - info->rect_o.y_min + 2 * bord);
+                Bitmap *dst = info->bm_o;
+                if (dst) {
+                    dst->left = info->rect_o.x_min - info->x - bord;
+                    dst->top  = info->rect_o.y_min - info->y - bord;
+                    for (int j = 0; j < info->bitmap_count; ++j) {
+                        Bitmap *src = info->bitmaps[j].image->bm_o;
+                        if (!src)
+                            continue;
+                        int x = info->bitmaps[j].x + src->left - dst->left;
+                        int y = info->bitmaps[j].y + src->top  - dst->top;
+                        assert(x >= 0 && x + src->w <= dst->w);
+                        assert(y >= 0 && y + src->h <= dst->h);
+                        unsigned char *buf = dst->buffer + y * dst->stride + x;
+                        render_priv->add_bitmaps_func(buf, dst->stride,
+                                                      src->buffer, src->stride,
+                                                      src->h, src->w);
                     }
                 }
             }
-            cleanup:
-            info = info->next;
         }
     }
 
-    for (i = 0; i < nb_bitmaps; ++i) {
-        if(!combined_info[i].cached){
+    for (int i = 0; i < nb_bitmaps; ++i) {
+        if (combined_info[i].bitmaps) {  // not cached
```

Comment:
Shouldn’t this be done in the same loop? Doing it in two separate loops can produce duplicates, can’t it?